hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/TruncateRegionProcedure.java

```diff
@@ -0,0 +1,212 @@
+        case TRUNCATE_REGION_MAKE_ONLINE:
+          addChildProcedure(createAssignProcedures(env));
+          setNextState(TruncateRegionState.TRUNCATE_REGION_POST_OPERATION);
+          break;
+        case TRUNCATE_REGION_POST_OPERATION:
+          postTruncate(env);
+          LOG.debug("truncate '" + getTableName() + "' completed");
+          return Flow.NO_MORE_STATE;
+        default:
+          throw new UnsupportedOperationException("unhandled state=" + state);
+      }
+    } catch (IOException e) {
+      if (isRollbackSupported(state)) {
+        setFailure("master-truncate-region", e);
+      } else {
+        LOG.warn("Retriable error trying to truncate region=" + getRegion().getRegionNameAsString()
+          + " state=" + state, e);
+      }
+    }
+    return Flow.HAS_MORE_STATE;
+  }
+
+  private void deleteRegionFromFileSystem(final MasterProcedureEnv env) throws IOException {
+    RegionStateNode regionNode =
+      env.getAssignmentManager().getRegionStates().getRegionStateNode(getRegion());
+    try {
+      regionNode.lock();
+      final MasterFileSystem mfs = env.getMasterServices().getMasterFileSystem();
+      final Path tableDir = CommonFSUtils.getTableDir(mfs.getRootDir(), getTableName());
+      HRegionFileSystem.deleteRegionFromFileSystem(env.getMasterConfiguration(),
+        mfs.getFileSystem(), tableDir, getRegion());
+    } finally {
+      regionNode.unlock();
+    }
+  }
+
+  @Override
+  protected void rollbackState(final MasterProcedureEnv env, final TruncateRegionState state)
+    throws IOException {
+    if (state == TruncateRegionState.TRUNCATE_REGION_PRE_OPERATION) {
+      // Nothing to rollback, pre-truncate is just table-state checks.
+      return;
+    }
+    if (state == TruncateRegionState.TRUNCATE_REGION_MAKE_OFFLINE) {
+      RegionStateNode regionNode =
+        env.getAssignmentManager().getRegionStates().getRegionStateNode(getRegion());
+      if (regionNode == null) {
+        // Region was unassigned by state TRUNCATE_REGION_MAKE_OFFLINE.
+        // So Assign it back
+        addChildProcedure(createAssignProcedures(env));
+      }
+      return;
+    }
+    // The truncate doesn't have a rollback. The execution will succeed, at some point.
+    throw new UnsupportedOperationException("unhandled state=" + state);
+  }
+
+  @Override
+  protected void completionCleanup(final MasterProcedureEnv env) {
+    releaseSyncLatch();
+  }
+
+  @Override
+  protected boolean isRollbackSupported(final TruncateRegionState state) {
+    return state == TruncateRegionState.TRUNCATE_REGION_PRE_OPERATION;
+  }
+
+  @Override
+  protected TruncateRegionState getState(final int stateId) {
+    return TruncateRegionState.forNumber(stateId);
+  }
+
+  @Override
+  protected int getStateId(final TruncateRegionState state) {
+    return state.getNumber();
+  }
+
+  @Override
+  protected TruncateRegionState getInitialState() {
+    return TruncateRegionState.TRUNCATE_REGION_PRE_OPERATION;
+  }
+
+  @Override
```

Comment:
Why was serializeStateData and deserializeStateData dropped? We need that right, at least super call?  
Also, Is user being serialized/deserialized somewhere?
 Also I prefer having the state declaration in proto for extensibility in future which was also dropped.

Reply:
> Why was serializeStateData and deserializeStateData dropped? We need that right, at least super call? Also, Is user being serialized/deserialized somewhere? Also I prefer having the state declaration in proto for extensibility in future which was also dropped.
Hi @NihalJain ,
I checked this aspect and have following observations.
- AbstractStateMachineRegionProcedure currently has 8 inheritors, out of which only 5 classes overrides serialise and deserialise because they have extra fields. So looks like general pattern is to override it only when its necessary. 
- Super class methods automatically gets and serialise/deserialise "region", even if we don't override in TruncateRegion. I verified it using debugger.
- User is not serialised, instead its populated in AbstractStateMachineTableProcedure constructor from the MasterProcedureEnv.
- At moment we don't foresee extra parameters for truncate_region. So whenever we get that requirements we can always add state proto and override the serialise/deserialise methods.

So based on these findings, I feel its okay to rely on superclass implementation.

Reply:
>> User is not serialised, instead its populated in AbstractStateMachineTableProcedure constructor from the MasterProcedureEnv.

I am not sure if it's not deserialized, how would it be loaded during failure. Please check code and see if contructor is invoked during failure. If not I would suggest to have this field serialzed/ deserialized to avoid having a null user or possibly a NPE during execution.